python312Packages.pygls: unable to build

### Nixpkgs version

- Unstable (25.11)

### Describe the bug

```
❯ nix-build -A python312Packages.pygls
this derivation will be built:
  /nix/store/8xnhhbfjqjqnbmz99cgzkysbbfqj847k-python3.12-pygls-1.3.1.drv
these 27 paths will be fetched (44.13 MiB download, 134.73 MiB unpacked):
  /nix/store/66qbb4n5vpdb6cps9f4nf61497s5373h-pypa-build-hook.sh
  /nix/store/49g1l3f8pls29x84b1hlsfw1lsn13ghc-pypa-install-hook
  /nix/store/laspr3dnbk0gwvbriwj2by157f2dw8m9-pytest-check-hook
  /nix/store/h85ci4zifnlc1iaz349hm36cns6aqcxb-python-catch-conflicts-hook
  /nix/store/8yiywxnv09bihzrg8mjaahj20nh7h6ir-python-imports-check-hook.sh
  /nix/store/c6yx6j4yj9zqsag6wx8vmn78w9zxcwfi-python-namespaces-hook.sh
  /nix/store/adfw2fjfn9cbnr4mgknd23pb13rywsf2-python-relax-deps-hook
  /nix/store/ipkaydgcygpc3f46qwvxfsxvfhby3g77-python-remove-tests-dir-hook
  /nix/store/v5b7lgmqb1rfkwmi00s8wg1z33dwrybs-python-runtime-deps-check-hook.sh
  /nix/store/18l19gnl8dq75v6knghbcgkrikmpc0yv-python3-3.12.11
  /nix/store/nx8lk5px08fmdprc6k78nczdpcb439g0-python3.12-attrs-25.3.0
  /nix/store/p6a5dk46s8g2xqg69x824phz022mkirj-python3.12-build-1.3.0
  /nix/store/s5k3pqallfb2vbzjsss23qkqxg7s1ir6-python3.12-cattrs-25.1.1
  /nix/store/k2wx2hd32p8qw3hsji98c47iqxric4bq-python3.12-iniconfig-2.1.0
  /nix/store/fy80ps5i98v9slb1jpn83d9s4p9bgrls-python3.12-installer-0.7.0
  /nix/store/bssscxd5w33v2k1m4yz2h3l7y77xm27f-python3.12-lsprotocol-2025.0.0
  /nix/store/mh3yf8wfj9n0xz3dl5bs26qgwhw8sn2s-python3.12-packaging-25.0
  /nix/store/93fbi42cynnbx2nd0rckiqkhj23bhks8-python3.12-pluggy-1.6.0
  /nix/store/cxy84w3pgi3k63gasm9lb0gkv4ginsv2-python3.12-poetry-core-2.1.3
  /nix/store/44gvjxxrvs6i2p5sff7lqi5hycjpq4ai-python3.12-pygments-2.19.2
  /nix/store/fchziciizl7dmxlh4xdzqhfp0giklq6y-python3.12-pyproject-hooks-1.2.0
  /nix/store/ggk45c8i53zyrs8w9vy7l4bqmmsjjjlc-python3.12-pytest-8.4.1
  /nix/store/pfncvqn6ysmlhphnk5dnzsvqx7ryz2gh-python3.12-pytest-asyncio-1.1.0
  /nix/store/49xhx546ld68wqffaypizjsx06ir079z-python3.12-typeguard-4.4.4
  /nix/store/rlp6vxhqlnzwffb65ab829xiwyl8dzpg-python3.12-typing-extensions-4.14.1
  /nix/store/208h4n4y0645w28y0z1whh8jff3nkv3j-python3.12-wheel-0.46.1
  /nix/store/cpmbn4pyzg065767x3rk6z9vn6fwzc45-wrap-python-hook
copying path '/nix/store/ipkaydgcygpc3f46qwvxfsxvfhby3g77-python-remove-tests-dir-hook' from 'https://cache.nixos.org'...
copying path '/nix/store/c6yx6j4yj9zqsag6wx8vmn78w9zxcwfi-python-namespaces-hook.sh' from 'https://cache.nixos.org'...
copying path '/nix/store/18l19gnl8dq75v6knghbcgkrikmpc0yv-python3-3.12.11' from 'https://cache.nixos.org'...
copying path '/nix/store/nx8lk5px08fmdprc6k78nczdpcb439g0-python3.12-attrs-25.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/h85ci4zifnlc1iaz349hm36cns6aqcxb-python-catch-conflicts-hook' from 'https://cache.nixos.org'...
copying path '/nix/store/8yiywxnv09bihzrg8mjaahj20nh7h6ir-python-imports-check-hook.sh' from 'https://cache.nixos.org'...
copying path '/nix/store/k2wx2hd32p8qw3hsji98c47iqxric4bq-python3.12-iniconfig-2.1.0' from 'https://cache.nixos.org'...
copying path '/nix/store/fy80ps5i98v9slb1jpn83d9s4p9bgrls-python3.12-installer-0.7.0' from 'https://cache.nixos.org'...
copying path '/nix/store/mh3yf8wfj9n0xz3dl5bs26qgwhw8sn2s-python3.12-packaging-25.0' from 'https://cache.nixos.org'...
copying path '/nix/store/93fbi42cynnbx2nd0rckiqkhj23bhks8-python3.12-pluggy-1.6.0' from 'https://cache.nixos.org'...
copying path '/nix/store/cxy84w3pgi3k63gasm9lb0gkv4ginsv2-python3.12-poetry-core-2.1.3' from 'https://cache.nixos.org'...
copying path '/nix/store/44gvjxxrvs6i2p5sff7lqi5hycjpq4ai-python3.12-pygments-2.19.2' from 'https://cache.nixos.org'...
copying path '/nix/store/fchziciizl7dmxlh4xdzqhfp0giklq6y-python3.12-pyproject-hooks-1.2.0' from 'https://cache.nixos.org'...
copying path '/nix/store/pfncvqn6ysmlhphnk5dnzsvqx7ryz2gh-python3.12-pytest-asyncio-1.1.0' from 'https://cache.nixos.org'...
copying path '/nix/store/rlp6vxhqlnzwffb65ab829xiwyl8dzpg-python3.12-typing-extensions-4.14.1' from 'https://cache.nixos.org'...
copying path '/nix/store/208h4n4y0645w28y0z1whh8jff3nkv3j-python3.12-wheel-0.46.1' from 'https://cache.nixos.org'...
copying path '/nix/store/cpmbn4pyzg065767x3rk6z9vn6fwzc45-wrap-python-hook' from 'https://cache.nixos.org'...
copying path '/nix/store/adfw2fjfn9cbnr4mgknd23pb13rywsf2-python-relax-deps-hook' from 'https://cache.nixos.org'...
copying path '/nix/store/49xhx546ld68wqffaypizjsx06ir079z-python3.12-typeguard-4.4.4' from 'https://cache.nixos.org'...
copying path '/nix/store/v5b7lgmqb1rfkwmi00s8wg1z33dwrybs-python-runtime-deps-check-hook.sh' from 'https://cache.nixos.org'...
copying path '/nix/store/p6a5dk46s8g2xqg69x824phz022mkirj-python3.12-build-1.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/s5k3pqallfb2vbzjsss23qkqxg7s1ir6-python3.12-cattrs-25.1.1' from 'https://cache.nixos.org'...
copying path '/nix/store/49g1l3f8pls29x84b1hlsfw1lsn13ghc-pypa-install-hook' from 'https://cache.nixos.org'...
copying path '/nix/store/66qbb4n5vpdb6cps9f4nf61497s5373h-pypa-build-hook.sh' from 'https://cache.nixos.org'...
copying path '/nix/store/ggk45c8i53zyrs8w9vy7l4bqmmsjjjlc-python3.12-pytest-8.4.1' from 'https://cache.nixos.org'...
copying path '/nix/store/bssscxd5w33v2k1m4yz2h3l7y77xm27f-python3.12-lsprotocol-2025.0.0' from 'https://cache.nixos.org'...
copying path '/nix/store/laspr3dnbk0gwvbriwj2by157f2dw8m9-pytest-check-hook' from 'https://cache.nixos.org'...
building '/nix/store/8xnhhbfjqjqnbmz99cgzkysbbfqj847k-python3.12-pygls-1.3.1.drv'...
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing pypa-build-hook
Using pypaBuildPhase
Sourcing python-runtime-deps-check-hook
Using pythonRuntimeDepsCheckHook
Sourcing pypa-install-hook
Using pypaInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing pytest-check-hook
Using pytestCheckPhase
Running phase: unpackPhase
unpacking source archive /nix/store/hhf92yllfm1aqxvpcw1p8jvmqip5s786-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file "source/tests/test_workspace.py"
Running phase: patchPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: configurePhase
no configure script, doing nothing
Running phase: buildPhase
Executing pypaBuildPhase
Creating a wheel...
pypa build flags: --no-isolation --outdir dist/ --wheel
* Getting build dependencies for wheel...
* Building wheel...
Successfully built pygls-1.3.1-py3-none-any.whl
Finished creating a wheel...
/build/source/dist /build/source
Unpacking to: unpacked/pygls-1.3.1...OK
Repacking wheel as ./pygls-1.3.1-py3-none-any.whl...OK
/build/source
Finished executing pypaBuildPhase
Running phase: pythonRuntimeDepsCheckHook
Executing pythonRuntimeDepsCheck
Checking runtime dependencies for pygls-1.3.1-py3-none-any.whl
Finished executing pythonRuntimeDepsCheck
Running phase: installPhase
Executing pypaInstallPhase
Successfully installed pygls-1.3.1-py3-none-any.whl
Finished executing pypaInstallPhase
Running phase: pythonOutputDistPhase
Executing pythonOutputDistPhase
Finished executing pythonOutputDistPhase
Running phase: fixupPhase
shrinking RPATHs of ELF executables and libraries in /nix/store/vygy8k7xzvw928kg5cbqmnmyrr876kwk-python3.12-pygls-1.3.1
checking for references to /build/ in /nix/store/vygy8k7xzvw928kg5cbqmnmyrr876kwk-python3.12-pygls-1.3.1...
patching script interpreter paths in /nix/store/vygy8k7xzvw928kg5cbqmnmyrr876kwk-python3.12-pygls-1.3.1
stripping (with command strip and flags -S -p) in  /nix/store/vygy8k7xzvw928kg5cbqmnmyrr876kwk-python3.12-pygls-1.3.1/lib
shrinking RPATHs of ELF executables and libraries in /nix/store/4zqyp5v93g6qq75gc01a853s04lap48l-python3.12-pygls-1.3.1-dist
checking for references to /build/ in /nix/store/4zqyp5v93g6qq75gc01a853s04lap48l-python3.12-pygls-1.3.1-dist...
patching script interpreter paths in /nix/store/4zqyp5v93g6qq75gc01a853s04lap48l-python3.12-pygls-1.3.1-dist
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
Running phase: installCheckPhase
no Makefile or custom installCheckPhase, doing nothing
Running phase: pythonCatchConflictsPhase
Running phase: pythonRemoveBinBytecodePhase
Running phase: pythonImportsCheckPhase
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: pygls
Running phase: pytestCheckPhase
Executing pytestCheckPhase
pytest flags: -m pytest
ImportError while loading conftest '/build/source/tests/conftest.py'.
tests/conftest.py:27: in <module>
    from pygls.workspace import Workspace
pygls/workspace/__init__.py:6: in <module>
    from .workspace import Workspace
pygls/workspace/workspace.py:32: in <module>
    from pygls.workspace.text_document import TextDocument
pygls/workspace/text_document.py:37: in <module>
    class TextDocument(object):
pygls/workspace/text_document.py:76: in TextDocument
    self, change: types.TextDocumentContentChangeEvent_Type1
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   AttributeError: module 'lsprotocol.types' has no attribute 'TextDocumentContentChangeEvent_Type1'
error: builder for '/nix/store/8xnhhbfjqjqnbmz99cgzkysbbfqj847k-python3.12-pygls-1.3.1.drv' failed with exit code 4;
       last 50 log lines:
       > Finished creating a wheel...
       > /build/source/dist /build/source
       > Unpacking to: unpacked/pygls-1.3.1...OK
       > Repacking wheel as ./pygls-1.3.1-py3-none-any.whl...OK
       > /build/source
       > Finished executing pypaBuildPhase
       > Running phase: pythonRuntimeDepsCheckHook
       > Executing pythonRuntimeDepsCheck
       > Checking runtime dependencies for pygls-1.3.1-py3-none-any.whl
       > Finished executing pythonRuntimeDepsCheck
       > Running phase: installPhase
       > Executing pypaInstallPhase
       > Successfully installed pygls-1.3.1-py3-none-any.whl
       > Finished executing pypaInstallPhase
       > Running phase: pythonOutputDistPhase
       > Executing pythonOutputDistPhase
       > Finished executing pythonOutputDistPhase
       > Running phase: fixupPhase
       > shrinking RPATHs of ELF executables and libraries in /nix/store/vygy8k7xzvw928kg5cbqmnmyrr876kwk-python3.12-pygls-1.3.1
       > checking for references to /build/ in /nix/store/vygy8k7xzvw928kg5cbqmnmyrr876kwk-python3.12-pygls-1.3.1...
       > patching script interpreter paths in /nix/store/vygy8k7xzvw928kg5cbqmnmyrr876kwk-python3.12-pygls-1.3.1
       > stripping (with command strip and flags -S -p) in  /nix/store/vygy8k7xzvw928kg5cbqmnmyrr876kwk-python3.12-pygls-1.3.1/lib
       > shrinking RPATHs of ELF executables and libraries in /nix/store/4zqyp5v93g6qq75gc01a853s04lap48l-python3.12-pygls-1.3.1-dist
       > checking for references to /build/ in /nix/store/4zqyp5v93g6qq75gc01a853s04lap48l-python3.12-pygls-1.3.1-dist...
       > patching script interpreter paths in /nix/store/4zqyp5v93g6qq75gc01a853s04lap48l-python3.12-pygls-1.3.1-dist
       > Executing pythonRemoveTestsDir
       > Finished executing pythonRemoveTestsDir
       > Running phase: installCheckPhase
       > no Makefile or custom installCheckPhase, doing nothing
       > Running phase: pythonCatchConflictsPhase
       > Running phase: pythonRemoveBinBytecodePhase
       > Running phase: pythonImportsCheckPhase
       > Executing pythonImportsCheckPhase
       > Check whether the following modules can be imported: pygls
       > Running phase: pytestCheckPhase
       > Executing pytestCheckPhase
       > pytest flags: -m pytest
       > ImportError while loading conftest '/build/source/tests/conftest.py'.
       > tests/conftest.py:27: in <module>
       >     from pygls.workspace import Workspace
       > pygls/workspace/__init__.py:6: in <module>
       >     from .workspace import Workspace
       > pygls/workspace/workspace.py:32: in <module>
       >     from pygls.workspace.text_document import TextDocument
       > pygls/workspace/text_document.py:37: in <module>
       >     class TextDocument(object):
       > pygls/workspace/text_document.py:76: in TextDocument
       >     self, change: types.TextDocumentContentChangeEvent_Type1
       >                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       > E   AttributeError: module 'lsprotocol.types' has no attribute 'TextDocumentContentChangeEvent_Type1'
       For full logs, run:
         nix log /nix/store/8xnhhbfjqjqnbmz99cgzkysbbfqj847k-python3.12-pygls-1.3.1.drv
```

### Steps to reproduce

nix-build -A python312Packages.pygls


### Expected behaviour

It should build successfully

### Screenshots

_No response_

### Relevant log output

```console

```

### Additional context

_No response_

### System metadata

```
❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.16.1, NixOS, 25.11 (Xantusia), 25.11.20250819.2007595`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.4`
 - channels(pol): `"nixos"`
 - nixpkgs: `/nix/store/pxnx7dbr0pz3qpgw2r315wdbb00afdw7-source`
```

### Notify maintainers



---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+bug%22+-label%3A%226.topic%3A+darwin%22+-label%3A%226.topic%3A+nixos%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/lsprotocol/2023.nix
+++ b/pkgs/development/python-modules/lsprotocol/2023.nix
@@ -9,26 +9,21 @@
   jsonschema,
   pyhamcrest,
   pytestCheckHook,
-  pythonOlder,
 }:
 
 buildPythonPackage rec {
   pname = "lsprotocol";
-  version = "2025.0.0";
+  version = "2023.0.1";
   pyproject = true;
-
-  disabled = pythonOlder "3.7";
 
   src = fetchFromGitHub {
     owner = "microsoft";
     repo = "lsprotocol";
     tag = version;
-    hash = "sha256-DrWXHMgDZSQQ6vsmorThMrUTX3UQU+DajSEOdxoXrFQ=";
+    hash = "sha256-PHjLKazMaT6W4Lve1xNxm6hEwqE3Lr2m5L7Q03fqb68=";
   };
 
-  postPatch = ''
-    pushd packages/python
-  '';
+  sourceRoot = "${src.name}/packages/python";
 
   build-system = [
     flit-core
@@ -48,21 +43,26 @@ buildPythonPackage rec {
   ];
 
   disabledTests = [
+    # cattrs.errors.StructureHandlerNotFoundError: Unsupported type:
+    # typing.Union[str, lsprotocol.types.NotebookDocumentFilter_Type1,
+    # lsprotocol.types.NotebookDocumentFilter_Type2,
+    # lsprotocol.types.NotebookDocumentFilter_Type3, NoneType]. Register
+    # a structure hook for it.
     "test_notebook_sync_options"
   ];
 
   preCheck = ''
-    popd
+    cd ../../
   '';
 
   pythonImportsCheck = [ "lsprotocol" ];
 
-  meta = with lib; {
+  meta = {
     description = "Python implementation of the Language Server Protocol";
     homepage = "https://github.com/microsoft/lsprotocol";
     changelog = "https://github.com/microsoft/lsprotocol/releases/tag/${src.tag}";
-    license = licenses.mit;
-    maintainers = with maintainers; [
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [
       doronbehar
       fab
     ];

--- a/pkgs/development/python-modules/lsprotocol/2025.nix
+++ b/pkgs/development/python-modules/lsprotocol/2025.nix
@@ -1,0 +1,70 @@
+{
+  lib,
+  attrs,
+  buildPythonPackage,
+  cattrs,
+  fetchFromGitHub,
+  flit-core,
+  importlib-resources,
+  jsonschema,
+  pyhamcrest,
+  pytestCheckHook,
+}:
+
+buildPythonPackage rec {
+  pname = "lsprotocol";
+  version = "2025.0.0";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "microsoft";
+    repo = "lsprotocol";
+    tag = version;
+    hash = "sha256-DrWXHMgDZSQQ6vsmorThMrUTX3UQU+DajSEOdxoXrFQ=";
+  };
+
+  sourceRoot = "${src.name}/packages/python";
+
+  build-system = [
+    flit-core
+  ];
+
+  dependencies = [
+    attrs
+    cattrs
+  ];
+
+  nativeCheckInputs = [ pytestCheckHook ];
+
+  checkInputs = [
+    importlib-resources
+    jsonschema
+    pyhamcrest
+  ];
+
+  disabledTests = [
+    # cattrs.errors.StructureHandlerNotFoundError: Unsupported type:
+    # typing.Union[str, lsprotocol.types.NotebookDocumentFilter_Type1,
+    # lsprotocol.types.NotebookDocumentFilter_Type2,
+    # lsprotocol.types.NotebookDocumentFilter_Type3, NoneType]. Register
+    # a structure hook for it.
+    "test_notebook_sync_options"
+  ];
+
+  preCheck = ''
+    cd ../../
+  '';
+
+  pythonImportsCheck = [ "lsprotocol" ];
+
+  meta = {
+    description = "Python implementation of the Language Server Protocol";
+    homepage = "https://github.com/microsoft/lsprotocol";
+    changelog = "https://github.com/microsoft/lsprotocol/releases/tag/${src.tag}";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [
+      doronbehar
+      fab
+    ];
+  };
+}

--- a/pkgs/development/python-modules/pygls/default.nix
+++ b/pkgs/development/python-modules/pygls/default.nix
@@ -10,6 +10,7 @@
   pythonOlder,
   typeguard,
   websockets,
+  nix-update-script,
 }:
 
 buildPythonPackage rec {
@@ -58,6 +59,14 @@ buildPythonPackage rec {
   '';
 
   pythonImportsCheck = [ "pygls" ];
+
+  passthru.updateScript = nix-update-script {
+    extraArgs = [
+      # Skips pre-releases
+      "--version-regex"
+      "^v([0-9.]+)$"
+    ];
+  };
 
   meta = with lib; {
     description = "Pythonic generic implementation of the Language Server Protocol";

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -8727,7 +8727,11 @@ self: super: with self; {
 
   lsp-tree-sitter = callPackage ../development/python-modules/lsp-tree-sitter { };
 
-  lsprotocol = callPackage ../development/python-modules/lsprotocol { };
+  lsprotocol = lsprotocol_2023;
+
+  lsprotocol_2023 = callPackage ../development/python-modules/lsprotocol/2023.nix { };
+
+  lsprotocol_2025 = callPackage ../development/python-modules/lsprotocol/2025.nix { };
 
   ltpycld2 = callPackage ../development/python-modules/ltpycld2 { };
 

